python/dgl/nn/pytorch/hetero.py

```diff
@@ -250,3 +251,137 @@ def get_aggregate_fn(agg):
         return _stack_agg_func
     else:
         return partial(_agg_func, fn=fn)
+
+class HeteroLinearLayer(nn.Module):
+    """Apply a linear transformation on the node features of a
+    heterogeneous graph and return a homogeneous graph representation.
+
+    The underlying implementation invokes dgl.to_homogeneous. Therefore,
+    the returned graph stores the following extra attributes:
+    
+      - ret_g.ndata[dgl.NID]: the original node IDs.
+      - ret_g.edata[dgl.EID]: the original edge IDs.
+      - ret_g.ndata[dgl.NTYPE]: the type ID of each node.
+      - ret_g.edata[dgl.ETYPE]: the type ID of each edge.
+
+    Parameters
+    ----------
+    hg : DGLGraph
+        A heterogeneous DGL graph.
+    out_size : int
+        Output feature size.
+    feat_name : str
+        Which feature to transform.
+
+    Examples
+    --------
+    >>> hg = dgl.heterograph({('user', 'rate', 'movie') : ...,
+                              ('user', 'follows', 'user') : ...})
+    >>> print(hg.num_nodes('user'), hg.num_nodes('movie'))
+    300 1000
+    >>> hg.nodes['user'].data['feat'] = torch.randn(hg.num_nodes('user'), 32)
+    >>> hg.nodes['movie'].data['feat'] = torch.randn(hg.num_nodes('movie'), 64)
+    >>> layer = dgl.nn.HeteroLinearLayer(hg, 100, 'feat')
+    >>> g, feat = layer(hg)
+    >>> print(g.num_nodes())
+    1300
+    >>> print(feat.shape)
+    (1300, 100)
+    """
+    def __init__(self, hg, out_size, feat_name):
+        super(HeteroLinearLayer, self).__init__()
+        self.feat_name = feat_name
+        self.linears = nn.ModuleDict()
+        for ntype in hg.ntypes:
+            linear = nn.Linear(hg.nodes[ntype].data[feat_name].shape[1], out_size)
+            self.linears[ntype] = linear
+    
+    def forward(self, hg):
+        """Forward function
+        
+        Parameters
+        ----------
+        hg : DGLGraph
+            The input heterogeneous DGL graph.
+
+        Returns
+        -------
+        g : DGLGraph
+            The homogenized DGL graph.
+        feat : Tensor
+            Transformed node features.
+        """
+        feat = th.tensor([])
+        for ntype in hg.ntypes:
+            features = self.linears[ntype](hg.nodes[ntype].data[self.feat_name])
+            feat = th.cat((feat, features))
+        g = to_homogeneous(hg)   
```

Comment:
Have you checked if the order of the nodes/node types in the homogenized graph is the same as that in the homogenized features by examining `to_homogeneous`?